core/src/cgroup.rs

```diff
@@ -185,16 +188,30 @@ impl CGroup {
         // Calling kill will also kill all sub cgroup processes
         self.kill()?;
 
-        // Obtain all sub cgroups
-        let dirs: Vec<PathBuf> = fs::read_dir(&self.path)?
-            .filter(|entry| entry.as_ref().unwrap().metadata().unwrap().is_dir())
-            .map(|entry| entry.unwrap().path())
-            .collect();
-
-        for dir in dirs {
-            fs::remove_dir(dir)?;
+        // Delete all cgroups from deepest to highest.
+        // It is necessary to delete the outer-most directories first, because
+        // a non-empty directory may not be deleted.
+        trace!("Calling remove on {}", &self.path.display());
+        for d in WalkDir::new(&self.path)
+            .into_iter()
+            .flatten()
+            .filter(|e| e.file_type().is_dir())
+            .sorted_by(|a, b| a.depth().cmp(&b.depth()).reverse())
+        {
+            trace!("Removing cgroup {}", &d.path().display());
+
+            // Re-try on every EBUSY
+            loop {
```

Comment:
Is there maybe a child process we could wait for instead of busy waiting?

Reply:
It could be possible figuring it out using `fuser`.

Reply:
We should be able to tell which processes we started and which resources we assigned to them. Therefore, we should already know which process we have to wait for without using `fuser`.

Reply:
> We should be able to tell which processes we started and which resources we assigned to them. Therefore, we should already know which process we have to wait for without using fuser.

I'm kinda curious about that. As far as I am aware of, there is no POSIX way of obtaining a list containing all alive child processes, meaning we would have to manage such a list ourselves, which is a very error-prone task. I'll take a look on how fuser works under the hood.

Reply:
`/proc/<PID>/task/<PID>/children` looks very promising.